File: Interviews/AV_Company/MaxScoreCircle.py
#coding=utf8

"""
思路：


"""

from math import sqrt

class Point(object):
    def __init__(self, x, y, score):
        self.x = x
        self.y = y
        self.score = score
        self.distance = self.x ** 2 + self.y ** 2

    #def __cmp__(self, other):
    #    return self.distance > other.distance

    def __repr__(self):
        return "(" + str(self.x) + ", " + str(self.y) + "): " + str(self.score)


class Solution(object):
    def maxScoreCircle(self, points):
        points.sort(key=lambda x: x.distance)
        last_radian_pow, curr_radian_pow = 0, 0
        score, accu_score = 0.0, 0.0
        for point in points:
            dist = point.distance
            if dist > curr_radian_pow:
                if accu_score > 1e-7:
                    score += accu_score
                    accu_score = 0.0
                    last_radian_pow = curr_radian_pow
                curr_radian_pow = dist
            accu_score += point.score

        if accu_score > 1e-7:
            score += accu_score
            last_radian_pow = curr_radian_pow
        return sqrt(last_radian_pow)

    def maxScoreCircle2(self, points):
        points.sort(key=lambda x: x.distance)
        score = 0
        scores = {}
        for pt in points:
            score += pt.score
            scores[pt.distance] = score
        return max(map(lambda x: x[1], scores.items()))

if __name__ == '__main__':
    s = Solution()
    pa = Point(1, 2, 3.4)
    pb = Point(-1, 3, 5.6)
    pc = Point(0, 5, 2.1)
    points = [pa, pb, pc]
    print "radian: ", s.maxScoreCircle(points)

File: Interviews/AV_Company/MaxScoreCircle.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

struct Point {
    Point(int x, int y, float score) {
        this->x = x;
        this->y = y;
        this->score = score;
        this->distance_sq = x * x + y * y;
    }
    
    bool operator <(const Point& rhs) const {
        return distance_sq < rhs.distance_sq;
    }

    std::string toString() {
        return "(" + to_string(x) + ", " + to_string(y) + ": " + to_string(score) + ")";
    }


    int x;
    int y;
    float score;
    int distance_sq;
};

std::ostream& operator<<(std::ostream &strm, const Point& p) {
    return strm << "(" << p.x << ", " << p.y << ": " << p.score  << ")";
}

void printPointVec(const vector<Point>& points) {
    for (size_t i = 0; i < points.size(); ++i) {
        cout << points[i] << " ";
    }
    cout << endl;
}

void maxScoreCircle(vector<Point>& points, float* radian, float* score) {
    std::sort(points.begin(), points.end());
    //printPointVec(points);
    float accu_score = 0.0; 
    int curr_radian = 0; 
    for (size_t i = 0; i < points.size(); ++i) {
        int distance = points[i].distance_sq;
        if (distance > curr_radian) {
            if (accu_score > 1e-7) {
                *score += accu_score;
                accu_score = 0;
                *radian = curr_radian;
            }
            curr_radian = distance; 
        }
        accu_score += points[i].score;
    }

    if (accu_score > 1e-7) {
        *score += accu_score;
        *radian = curr_radian;
    }
}



int main() {
    vector<Point> points;
    points.push_back(Point(1, 2, 1.2));
    points.push_back(Point(0, 0, 2.5));

    float radian = 0.0;
    float score = 0.0;
    maxScoreCircle(points, &radian, &score);
    cout << "Radian: " << radian << endl;
    cout << "Score: " << score << endl;

    return 0;
}
